src/stream/src/executor/source/mod.rs

```diff
@@ -134,30 +136,67 @@ pub async fn apply_rate_limit(stream: BoxSourceChunkStream, rate_limit_rps: Opti
     #[for_await]
     for chunk in stream {
         let chunk = chunk?;
-        let chunk_size = chunk.capacity();
+        yield process_chunk(chunk, rate_limit_rps, &limiter).await;
+    }
+}
 
-        if rate_limit_rps.is_none() || chunk_size == 0 {
-            // no limit, or empty chunk
-            yield chunk;
-            continue;
-        }
+#[try_stream(ok = Option<StreamChunk>, error = ConnectorError)]
+pub async fn apply_rate_limit_with_for_streaming_file_source_reader(
+    stream: BoxStreamingFileSourceChunkStream,
+    rate_limit_rps: Option<u32>,
+) {
```

Comment:
Do you mean to avoid nesting try_stream as much as we can? @kwannoel 
The upper layer `apply_rate_limit` can be removed?

Reply:
Not related.